File: erlang/seminars/barber/waiting.erl
-module(waiting).

-export([start/1]).


start(Max) ->
    {ok, spawn_link(fun() -> init(Max) end)}.

init(Max) ->
    waiting(Max, []).

waiting(0, []) ->
    io:format("should not happen ~n", []),
    ok;
waiting(0, [Next|Rest]=Queue) ->
    receive 
	close ->
	    ok;
	{enter, Customer} ->
	    Customer ! sorry,
	    waiting(0, Queue);
	{next, Barber} ->
	    Barber ! {ok, Next},
	    waiting(1, Rest)
    end;
waiting(N, []) ->
    receive
	close ->
	    ok;
	{enter, Customer} ->
	    Customer ! please_wait,
	    waiting(N-1, [Customer])
    end;
waiting(N, [Next|Rest]=Queue) ->
    receive
	close ->
	    ok;
	{next, Barber} ->
	    Barber ! {ok, Next},
	    waiting(N+1, Rest);
	{enter, Customer} ->
	    Customer ! please_wait,
	    waiting(N-1, Queue ++ [Customer])
    end.

File: exercises/barber/src/waiting.ex
defmodule Waiting do
  
  def start(max) do
    {:ok, spawn_link(fn -> init(max) end)}
  end

  defp init(max), do: waiting(max, [])

  # Manages the waiting system of the barber shop including
  # the waiting times and the logic of customers entering
  # the shop and the barber's work. Gently stop the process
  # when the shop closes.
  defp waiting(0, []) do
    IO.puts("Should not happen!")
    :ok
  end
  defp waiting(0, [next | rest] = queue) do
    receive do
      :close ->
        :ok

      {:enter, customer} ->
        send(customer, :sorry)
        waiting(0, queue)

      {:next, barber} ->
        send(barber, {:ok, next})
        waiting(1, rest)
    end
  end
  defp waiting(n, []) do
    receive do
      :close ->
        :ok

      {:enter, customer} ->
        send(customer, :please_wait)
        waiting(n - 1, [customer])
    end
  end
  defp waiting(n, [next | rest] = queue) do
    receive do
      :close ->
        :ok

      {:next, barber} ->
        send(barber, {:ok, next})
        waiting(n + 1, rest)

      {:enter, customer} ->
        send(customer, :please_wait)
        waiting(n - 1, queue ++ [customer])
    end
  end
end
